### ISSUE TITLE
 NoMethodError: undefined method `resolv_nbo' for Socket:Class when using a proxy such setg Proxies socks4:x.x.x.x:31337

### ISSUE THREAD

**Issue by pyro57** [rapid7/metasploit-framework#17981]
## Steps to reproduce

How'd you do it?

1. spin up a sock4 proxy such as through cobalt strike
2. run setg Proxies IP:Port within metasploit
3. attempt to run a module such as auxilary/scanner/smb/smb_ms17_010

This section should also tell us any relevant information about the
environment; for example, if an exploit that used to work is failing,
tell us the victim operating system and service versions.

## Were you following a specific guide/tutorial or reading documentation?

I was not following a guide, but conducting a penetration test.

## Expected behavior

the traffic should flow through the proxy and hit the target.

## Current behavior

the module errors out with the following error message
- NoMethodError: undefined method `resolv_nbo' for Socket:Class

### Metasploit version
6.3.15-dev

##  Module/Datastore

The following global/module datastore, and database setup was configured before the issue occurred:
<details>
<summary>Collapse</summary>

```
[framework/core]
Proxies=socks4:x.x.x.x:31337

[framework/ui/console]
ActiveModule=auxiliary/scanner/smb/smb_ms17_010

[scanner/smb/smb_ms17_010]
RHOSTS=x.x.x.x
loglevel=3
WORKSPACE=
VERBOSE=false
RPORT=445
SSL=false
SSLServerNameIndication=
SSLVersion=Auto
SSLVerifyMode=PEER
SSLCipher=
Proxies=socks4:x.x.x.x:31337
CPORT=
CHOST=
ConnectTimeout=10
TCP::max_send_size=0
TCP::send_delay=0
DCERPC::max_frag_size=4096
DCERPC::fake_bind_multi=true
DCERPC::fake_bind_multi_prepend=0
DCERPC::fake_bind_multi_append=0
DCERPC::smb_pipeio=rw
DCERPC::ReadTimeout=10
NTLM::UseNTLMv2=true
NTLM::UseNTLM2_session=true
NTLM::SendLM=true
NTLM::UseLMKey=false
NTLM::SendNTLM=true
NTLM::SendSPN=true
SMB::pipe_evasion=false
SMB::pipe_write_min_size=1
SMB::pipe_write_max_size=1024
SMB::pipe_read_min_size=1
SMB::pipe_read_max_size=1024
SMB::pad_data_level=0
SMB::pad_file_level=0
SMB::obscure_trans_pipe_level=0
SMBDirect=true
SMBUser=
SMBPass=
SMBDomain=.
SMBName=*SMBSERVER
SMB::VerifySignature=false
SMB::ChunkSize=500
SMB::Native_OS=Windows 2000 2195
SMB::Native_LM=Windows 2000 5.0
SMB::ProtocolVersion=1,2,3
SMB::AlwaysEncrypt=true
KrbCacheMode=read-write
SMB::Auth=auto
SMB::Rhostname=
DomainControllerRhost=
SMB::Krb5Ccname=
SMB::KrbOfferedEncryptionTypes=AES256,AES128,RC4-HMAC,DES-CBC-MD5,DES3-CBC-SHA1
NAMED_PIPES=/opt/metasploit/data/wordlists/named_pipes.txt
THREADS=1
ShowProgress=true
ShowProgressPercent=10
CHECK_DOPU=true
CHECK_ARCH=true
CHECK_PIPE=false
```

</details>

##  Database Configuration

The database contains the following information:
<details>
<summary>Collapse</summary>

```
Session Type: Connected to msf. Connection type: postgresql.
```

| ID | Hosts | Vulnerabilities | Notes | Services |
|-:|-:|-:|-:|-:|
| 1 **(Current)** | 0 | 0 | 0 | 0 |
| **Total (1)** | **0** | **0** | **0** | **0** |

</details>

##  History

The following commands were ran during the session and before this issue occurred:
<details>
<summary>Collapse</summary>

```
203    setg Proxies socks4:x.x.x.x:31337
204    use auxiliary/scanner/smb/smb_ms17_010
205    set RHOSTS x.x.x.x
206    run
207    version
208    set loglevel 3
209    run
210    debug
```

</details>

##  Framework Errors

The following framework errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:40:57] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:40:57] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:40:57] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[05/10/2023 13:16:06] [e(0)] core: DB.connect threw an exception - ActiveRecord::ConnectionNotEstablished connection to server at "127.0.0.1", port 5433 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?

[05/10/2023 13:16:06] [e(0)] core: Failed to connect to the database: connection to server at "127.0.0.1", port 5433 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
```

</details>

##  Web Service Errors

The following web service errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
msf-ws.log does not exist.
```

</details>

##  Framework Logs

The following framework logs were recorded before the issue occurred:
<details>
<summary>Collapse</summary>

```
[04/27/2023 09:38:21] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:38:46] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:53] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:39:54] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:40:57] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:40:57] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 09:40:57] [e(0)] core: NoMethodError undefined method `resolv_nbo' for Socket:Class
[04/27/2023 13:26:41] [d(0)] core: HistoryManager.pop_context name: :msfconsole
[05/10/2023 13:16:06] [i(0)] core: Trying to continue despite failed database creation: connection to server at "127.0.0.1", port 5433 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?

[05/10/2023 13:16:06] [e(0)] core: DB.connect threw an exception - ActiveRecord::ConnectionNotEstablished connection to server at "127.0.0.1", port 5433 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?

[05/10/2023 13:16:06] [e(0)] core: Failed to connect to the database: connection to server at "127.0.0.1", port 5433 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?

[05/10/2023 13:16:06] [d(0)] core: Updated user based module store
[05/10/2023 13:17:12] [d(0)] core: HistoryManager.push_context name: :msfconsole
[05/10/2023 13:33:06] [d(0)] core: HistoryManager.pop_context name: :msfconsole
[05/10/2023 13:36:31] [d(0)] core: HistoryManager.push_context name: :msfconsole
[05/10/2023 13:39:16] [d(0)] core: HistoryManager.pop_context name: :msfconsole
[05/10/2023 14:39:43] [d(0)] core: HistoryManager.push_context name: :msfconsole
```

</details>

##  Web Service Logs

The following web service logs were recorded before the issue occurred:
<details>
<summary>Collapse</summary>

```
msf-ws.log does not exist.
```

</details>

##  Version/Install

The versions and install method of your Metasploit setup:
<details>
<summary>Collapse</summary>

```
Framework: 6.3.15-dev
Ruby: ruby 3.0.5p211 (2022-11-24 revision ba5cf0f7c5) [x86_64-linux]
OpenSSL: OpenSSL 3.0.7 1 Nov 2022
Install Root: /opt/metasploit
Session Type: Connected to msf. Connection type: postgresql.
Install Method: Other - Please specify
```

</details>

**Comment by adfoster-r7** [comment]
Pull request to fix - https://github.com/rapid7/rex-socket/pull/60

A work around is applying that fix locally, or using socks5 instead of socks4. Or wait for it to be shipped in the next weekly release

**Comment by SolidCunno** [comment]
i just had the same issue. i am learning pivot pentesting and the same thing happened happened.

**Comment by adfoster-r7** [comment]
Should be fixed in version 6.3.17
